internal/store/device.go

```diff
@@ -181,88 +188,105 @@ func (s *DeviceStore) Get(ctx context.Context, orgId uuid.UUID, name string) (*a
 	return &apiDevice, nil
 }
 
-func (s *DeviceStore) CreateOrUpdate(ctx context.Context, orgId uuid.UUID, resource *api.Device, fieldsToUnset []string, fromAPI bool, callback DeviceStoreCallback) (*api.Device, bool, error) {
+func (s *DeviceStore) getExistingRecord(name string, orgId uuid.UUID) (*model.Device, bool, error) {
+	existingRecord := &model.Device{Resource: model.Resource{OrgID: orgId, Name: name}}
+	if err := s.db.First(existingRecord).Error; err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return nil, false, nil
+		}
+		return nil, false, flterrors.ErrorFromGormError(err)
+	}
+	return existingRecord, true, nil
+}
+
+func (s *DeviceStore) createDevice(device *model.Device) (bool, error) {
+	device.Generation = lo.ToPtr[int64](1)
+	device.ResourceVersion = lo.ToPtr[int64](1)
+	if result := s.db.Create(device); result.Error != nil {
+		err := flterrors.ErrorFromGormError(result.Error)
+		return err == flterrors.ErrDuplicateName, err
+	}
+	return false, nil
+}
+
+func (s *DeviceStore) updateDevice(fromAPI bool, existingRecord, device *model.Device, fieldsToUnset []string) (bool, error) {
+	if device.ResourceVersion != nil && lo.FromPtr(existingRecord.ResourceVersion) != lo.FromPtr(device.ResourceVersion) {
```

Comment:
Move this check down after the check for changing the spec of an owned device

Reply:
OK